src/databricks/labs/ucx/source_code/pyspark.py

```diff
@@ -1,10 +1,155 @@
 import ast
-from collections.abc import Iterable
+from abc import ABC, abstractmethod
+from collections.abc import Iterable, Iterator
+from dataclasses import dataclass
 
-from databricks.labs.ucx.source_code.base import Advice, Fixer, Linter
+from databricks.labs.ucx.source_code.base import Advice, Deprecation, Fixer, Linter
 from databricks.labs.ucx.source_code.queries import FromTable
 
 
+@dataclass
+class Matcher(ABC):
+    method_name: str
+    min_args: int
+    max_args: int
+    table_arg_index: int
+
+    def matches(self, node: ast.AST):
+        return (
+            isinstance(node, ast.Call)
+            and isinstance(node.func, ast.Attribute)
+            and self.min_args <= len(node.args) <= self.max_args
+            and isinstance(node.args[self.table_arg_index], ast.Constant)
+        )
+
+    @abstractmethod
+    def lint(self, from_table: FromTable, node: ast.Call) -> Iterator[Advice]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def apply(self, from_table: FromTable, node: ast.Call) -> None:
+        raise NotImplementedError()
+
+
+@dataclass
+class QueryMatcher(Matcher):
+
+    def lint(self, from_table: FromTable, node: ast.Call) -> Iterator[Advice]:
+        table_arg = node.args[self.table_arg_index]
+        assert isinstance(table_arg, ast.Constant)
```

Comment:
and if it is `ast.Name` or `ast.JoinedStr`? probably yielding a warning is better other than skipping it. We can make things more configurable as well.

Reply:
This is mimicked from the existing code ;-) but let's improve that indeed

Reply:
done